Implement per-user series limits

juliusv opened the pull request:
When a series limit is reached, samples for existing series should still
be stored, so this is treated as a partial failure.

I tried various approaches to error propagation and handling, including
more complicated ones, but went for this medium-simple approach in the
end that uses gRPC errors to communicate a series-exceeded error from
the ingester back to the distributor, still manages to only partially
fail for the affected series, and communicates the specific error back
to the user. Due to only returning the last error encountered when
writing to multiple ingesters, this may at times cover up the
series-exceeded error if there is another error afterwards.

juliusv commented on ingester/user_state.go:
```diff
@@ -165,13 +171,18 @@ func (u *userState) unlockedGet(metric model.Metric) (model.Fingerprint, *memory
+		return fp, series, nil
+	}
+
+	if u.fpToSeries.length() >= maxSeries {
```
There's theoretically a (relatively harmless) race here if multiple goroutines get the length of the series map at the same time, then all proceed to add a new series. I'm not sure if it's worth addressing, as this should rarely happen (all samples from one push are added serially), and the possible overshoot in allowed series would probably be minimal.

tomwilkie replied:
Pls stick this in a comment above this line.

tomwilkie commented:
Great!  We really need to also limit the number of series for a given metric name (as this is the easiest way to overwhelm the index) to something like 10 or 20k.  Worth doing in this PR or a different one?

tomwilkie commented on distributor/distributor.go:
```diff
@@ -330,6 +331,11 @@ func (d *Distributor) Push(ctx context.Context, req *remote.WriteRequest) (*cort
+				// We can't wrap this error or it won't be communicated back to
+				// the user correctly.
+				return nil, lastErr
+			}
```
In #272 I get rid of the custom error below and just propagate the error from the ingester (so compatible with this change).

I do need to give a bit of a think about how this throttling will interact with the quorum writes though...

tomwilkie replied:
> I do need to give a bit of a think about how this throttling will interact with the quorum writes though...

It'll be fine for now, I think...

tomwilkie commented on distributor/distributor.go:
```diff
@@ -361,6 +367,9 @@ func (d *Distributor) sendSamples(ctx context.Context, ingester *ring.IngesterDe
 		_, err := client.Push(ctx, util.ToWriteRequest(samples))
+		if grpc.Code(err) == codes.ResourceExhausted && grpc.ErrorDesc(err) == util.ErrUserSeriesLimitExceeded.Error() {
+			err = util.ErrUserSeriesLimitExceeded
+		}
```
I see what you mean; this is yucky.  Would it make sense to:
- propagate this error directly
- remove the "need n successful writes" error and propagate any error directly in that place?
- put all the error-mapping logic in http_server.go?

tomwilkie commented on ingester/user_state.go:
```diff
@@ -28,10 +29,11 @@ type userState struct {
 }
 
-func newUserStates(rateUpdatePeriod time.Duration) *userStates {
+func newUserStates(rateUpdatePeriod time.Duration, maxSeriesPerUser int) *userStates {
```
I'd probably propagated the whole of Config struct here, as we're already passing two things through.  Or better yet, make a new Config struct (`UserLimitsConfig`), embed that in Config and pass that here?

juliusv replied:
Hmm, `rateUpdatePeriod` is not a limit (it's only used to be able to show rates in the UI), so `UserLimitsConfig` doesn't fit so well. Maybe just `UserStatesConfig` since the only commonality is to configure user states? But then at two arguments, I think it makes more sense to keep it arguments?

tomwilkie replied:
+1 UserStatesConfig - I suspect we're going to have more shortly (to limit the number of series for a single metric name, for instance)

tomwilkie commented:
Couple of relatively minor things, and one extra ask.

juliusv commented:
@tomwilkie I did everything you remarked except for the per-metric limiting yet (will do separately). Separate commit for the `UserStatesConfig`, sorry, squashed the other feedback when I rebased.

tomwilkie commented:
Thanks!  LGTM
